### PR TITLE
[ADD] hr_cae: Applicant, Employee creation, Promotions

### PR DESCRIPTION
**Pull request by mclaeysb**
Also includes some extra fields in employee.

### CODE REVIEW

File: hr_cae/models/hr_applicant.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        string="Certificate Date", help="Certificate Delivery Date", required=False
+    )
+    professional_experience = fields.Text(
+        string="Professional Experience", required=False
+    )
+
+    @api.onchange("partner_id")
+    def onchange_partner_id(self):
+        res = super(Applicant, self).onchange_partner_id()
+        self.title = self.partner_id.title
```
Comment (robinkeunen) [inline hr_cae/models/hr_applicant.py:34]: 👍

File: hr_cae/models/hr_applicant.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class Applicant(models.Model):
+    _inherit = "hr.applicant"
+
+    title = fields.Many2one("res.partner.title")
+    job = fields.Char(string="Job", required=False)
+    job_description = fields.Text(string="Job Description", required=False)
+    equipment = fields.Text(string="Equipment", required=False)
+    type_id = fields.Many2one("hr.recruitment.degree", string="Certificate")
```
Comment (robinkeunen) [inline hr_cae/models/hr_applicant.py:19]: why not "degree_id"?

Comment (robinkeunen) [reply hr_cae/models/hr_applicant.py:19]: or `certificate_id`

File: hr_cae/models/hr_employee.py [hunk 3]
```diff
@@ -31,7 +32,13 @@ class Employee(models.Model):
     )
     certificate_id = fields.Many2one(
```
Comment (robinkeunen) [inline hr_cae/models/hr_employee.py:36]: why not `degree_id`?

File: hr_cae/tests/test_hr_cae.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+from odoo import fields
+from odoo.tests.common import TransactionCase
+
+
+class TestHRCAE(TransactionCase):
+    def test_create_employee_from_applicant(self):
+        applicant = self.browse_ref("hr_recruitment.hr_case_salesman0")
+        applicant.type_id = self.browse_ref("hr_recruitment.degree_graduate")
+        applicant.certificate_date = fields.Date.today()
+        applicant.professional_experience = "A lot of experience"
```
Comment (robinkeunen) [inline hr_cae/tests/test_hr_cae.py:14]: wow, le dude, engagez-le!

**Review by robinkeunen — APPROVED** [review]
Superficial naming comments, otherwise good work 👌

File: hr_cae/models/hr_applicant.py (hunk 2 above)
Comment (mclaeysb) [reply hr_cae/models/hr_applicant.py:19]: type_id already exists in `hr_applicant`. Here I'm just overwriting the string.

File: hr_cae/models/hr_employee.py (hunk 3 above)
Comment (mclaeysb) [reply hr_cae/models/hr_employee.py:36]: Here it's new.

**Comment by mclaeysb** [comment]
How come CI is not running anymore?

File: hr_cae/models/hr_employee.py [hunk 5]
```diff
@@ -10,12 +10,19 @@
...
+    title = fields.Many2one("res.partner.title")
     country_department_of_birth_id = fields.Many2one(
         "res.country.department", string="Department (France) of Birth", required=False
     )
     country_other_ids = fields.Many2many(
         "res.country", string="Other Nationalities", required=False
     )
+    promotion_id = fields.Many2one(
+        comodel_name="hr.promotion", string="Promotion", required=False
+    )
```
Comment (quentinDupont) [inline hr_cae/models/hr_employee.py]: Imho should be in custom repo because it's too specific :)

File: hr_cae/models/hr_employee.py [hunk 6]
```diff
@@ -31,14 +38,21 @@ class Employee(models.Model):
...
+        string="Professional Experience", required=False
+    )
     bank_account_payment_id = fields.Many2one(
         "res.partner.bank",
         string="Bank Account Number for Payment",
         required=False,
         domain="[('partner_id', '=', address_home_id)]",
         help="Employee bank salary account",
     )  # TODO: check that this one is actually used in logic
+    turnover_minimum = fields.Monetary(string="Minimum Turn-Over")
```
Comment (quentinDupont) [inline hr_cae/models/hr_employee.py:52]: Agreed, I'm not sure to understand the logic behind this field. Not blocking for me

File: hr_cae/models/hr_promotion.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class Promotion(models.Model):
+    _name = "hr.promotion"
+    _description = "Promotion"
+    _order = "date_start,state"
+
```
Comment (quentinDupont) [inline hr_cae/models/hr_promotion.py]: Yes, for me, it's should be in specific repo

**Review by quentinDupont — COMMENTED** [review]
My comments are about Promotions that's should be in other repo.
Actually, I also think that Recruitment could be separated. It's very interesting but, as as small cooperative that hire not that much, we're not gonna use it.
What do you think to create specific repos hr_cae_promotion, hr_cae_recruitement ?

File: hr_cae/models/hr_applicant.py (hunk 2 above)
Comment (robinkeunen) [reply hr_cae/models/hr_applicant.py:19]: I guess we decided to use type for certificate and that I forgot that I reluctantly agreed with it. :p

File: hr_cae/models/hr_promotion.py (hunk 7 above)
Comment (robinkeunen) [reply hr_cae/models/hr_promotion.py]: I tend to agree with @quentinDupont . What do you think @mclaeysb ?

File: hr_cae/models/hr_promotion.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            )
+            promotion.no_of_applicants = len(applicants)
+
+    @api.multi
+    @api.constrains("no_of_places_available")
+    def _constrain_no_of_places_available(self):
+        for promotion in self:
+            if promotion.no_of_places_available < 0 and promotion.no_of_places_max > 0:
+                raise ValidationError(
+                    _("Not enough places available in this promotion.")
```
Comment (robinkeunen) [inline hr_cae/models/hr_promotion.py]: I think "spot" is a better translation.

File: hr_cae/models/hr_promotion.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    @api.multi
+    def _compute_attachment_number(self):
+        read_group_res = self.env["ir.attachment"].read_group(
+            [("res_model", "=", "hr.promotion"), ("res_id", "in", self.ids)],
+            ["res_id"],
+            ["res_id"],
+        )
+        attach_data = {res["res_id"]: res["res_id_count"] for res in read_group_res}
+        for promotion in self:
+            promotion.attachment_number = attach_data.get(promotion.id, 0)
```
Comment (robinkeunen) [inline hr_cae/models/hr_promotion.py]: Nice use of read_group. I'd use a better name for `attach_data` though.

File: hr_cae/security/ir.model.access.csv [hunk 10]
```diff
@@ -19,3 +19,5 @@ access_hr_sector_user,access_hr_sector,model_hr_sector,hr.group_hr_user,1,0,0,0
 access_hr_sector_manager,access_hr_sector,model_hr_sector,hr.group_hr_manager,1,1,1,1
 access_hr_contribution_exemption_user,access_hr_contribution_exemption,model_hr_contribution_exemption,hr.group_hr_user,1,0,0,0
 access_hr_contribution_exemption_manager,access_hr_contribution_exemption,model_hr_contribution_exemption,hr.group_hr_manager,1,1,1,1
+access_hr_promotion_user,access_hr_promotion,model_hr_promotion,hr.group_hr_user,1,0,0,0
```
Comment (robinkeunen) [inline hr_cae/security/ir.model.access.csv]: I'm surprised users can't create promotions.

**Review by robinkeunen — COMMENTED** [review]
Just a few other minor comments.

Let's start the discussion with Quentin about putting promotions in another module though.

Also, it'd be great to include more detail in the PR description so Quentin and Sylvain can better understand the context and contribute easily. Just a copy/paste of the "functional description" of our specification would be enough I think.

File: hr_cae/models/hr_promotion.py (hunk 7 above)
Comment (robinkeunen) [reply hr_cae/models/hr_promotion.py]: I do agree for promotions but less so for recruitment. There would be no much left in hr_cae. What you use hr_cae modules for?

File: hr_cae/models/hr_employee.py (hunk 6 above)
Comment (mclaeysb) [reply hr_cae/models/hr_employee.py:52]: This is a field where CAE members can indicate the turnover they will try to make.

File: hr_cae/models/hr_promotion.py (hunk 7 above)
Comment (mclaeysb) [reply hr_cae/models/hr_promotion.py]: I share @robinkeunen's opinion. I separated Promotions!

File: hr_cae/models/hr_promotion.py (hunk 9 above)
Comment (mclaeysb) [reply hr_cae/models/hr_promotion.py]: This was a clean copy from our code in `vertical_housing_cooperative`. I'd be interested to know how to improve and standardise this.

File: hr_cae/security/ir.model.access.csv (hunk 10 above)
Comment (mclaeysb) [reply hr_cae/security/ir.model.access.csv]: Corrected!

**Comment by mclaeysb** [comment]
I've put Promotions in separate module, and did some other fixes and improvements!

**Pull request merged**